### PR TITLE
Fix indexing of empty array fields

### PR DESCRIPTION
**Pull request by alfinkel**
_What:_

Fix a bug that, incorrectly, does not index documents that contain an empty array field value.

_Why:_

Documents that contain an empty array should be index-able.

_How:_

Fix logic in the index updater that handles SQL INSERT statement creation for documents to be stored in Query index tables.  Logic for handling an empty array field value is now the same as the handling of an index-able field with a missing value in a document.

_Tests:_

Add two additional index updater test cases to the test empty array field values.

reviewer @mikerhodes 
reviewer @gadamc 

BugId: 48139

### CODE REVIEW

**Comment by gadamc** [comment]
Okay -- looks fine to me. +1

**Comment by alfinkel** [comment]
Thanks @gadamc but you're not listed as one of the reviewers. :)

File: Classes/common/query/CDTQIndexUpdater.m [hunk 1]
```diff
@@ -397,12 +396,14 @@ + (CDTQSqlParts *)createPartsForFieldNames:(NSArray *)fieldNames
             [CDTQValueExtractor extractValueForFieldName:fieldName fromDictionary:rev.body];
 
         if (value) {
+            // Do not include an empty array field as you would not include a field with no value.
+            if ([value isKindOfClass:[NSArray class]] && ((NSArray *)value).count == 0) {
+                continue;
+            }
+            
             [includedFieldNames addObject:fieldName];
             [args addObject:value];
```
Comment (gadamc) [inline Classes/common/query/CDTQIndexUpdater.m]: Same question from sync-android PR. Doesn't value get added to the args array twice?  And does that matter?

**Comment by gadamc** [comment]
This might be a separate PR -- but can we add a test for the case where an element of an indexed array is another object? As in pets: [cat, dog, {llama: bob}, fish] Does llama:bob get indexed and can we successfully query for it?

File: Classes/common/query/CDTQIndexUpdater.m (hunk 1 above)
Comment (alfinkel) [inline Classes/common/query/CDTQIndexUpdater.m]: Same answer as sync-android PR.  Take a look a few lines above at https://github.com/cloudant/CDTDatastore/pull/165/files#diff-5440a5114825a38f7d2feeefa446aabfR391.  This will skip fields and arguments included with initial settings.

Comment (gadamc) [inline Classes/common/query/CDTQIndexUpdater.m]: thanks

**Comment by gadamc** [comment]
barring your discussion with Mike about the tests for objects as elements of an indexed array, I'm +1 if you don't add that test to this PR.

**Comment by alfinkel** [comment]
Actually here is more info on the topic of sub-documents inside arrays:  `Dotted notation to index or query sub-documents in arrays.` is unsupported.  FB Case 47341 has been created to track this work.

**Comment by mikerhodes** [comment]
See my comments on logic structuring in https://github.com/cloudant/sync-android/pull/167.

**Comment by alfinkel** [comment]
Restructured `if` statements in fb3e03d

File: Classes/common/query/CDTQIndexUpdater.m [hunk 2]
```diff
@@ -331,47 +331,45 @@ + (CDTQSqlParts *)partsToDeleteIndexEntriesForDocId:(NSString *)docId
...
-
             CDTQSqlParts *parts = [CDTQIndexUpdater createPartsForFieldNames:fieldNames
-                                                       initialIncludedFields:includedFieldNames
-                                                         initialPlaceholders:placeholders
-                                                                 initialArgs:args
-                                                                   indexName:indexName
+                                                        initialIncludedFields:includedFieldNames
+                                                            initialPlaceholders:placeholders
+                                                                    initialArgs:args
+                                                                    indexName:indexName
```
Comment (mikerhodes) [inline Classes/common/query/CDTQIndexUpdater.m]: Spacing broken.

**Comment by mikerhodes** [comment]
+1 bar minor spacing issue.
